api/server/security/security_token.go

```diff
@@ -0,0 +1,151 @@
+package security
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	jcrypto "github.com/SermoDigital/jose/crypto"
+	"github.com/SermoDigital/jose/jws"
+
+	"github.com/codedellemc/libstorage/api/types"
+)
+
+var rxBearer = regexp.MustCompile(`Bearer (.+)`)
+
+// GetBearerTokenFromReq retrieves the bearer token from the HTTP request.
+func GetBearerTokenFromReq(ctx types.Context, req *http.Request) string {
+	m := rxBearer.FindStringSubmatch(
+		req.Header.Get(types.AuthorizationHeader))
+	if len(m) == 0 {
+		return ""
+	}
+	return m[1]
+}
+
+// ValidateSecurityToken validates a JWT and returns a SecurityToken if the
+// JWT is valid; otherwise a nil value is returned.
+func ValidateSecurityToken(
+	ctx types.Context,
+	config *types.AuthConfig,
+	encJWT string) (*types.SecurityToken, error) {
+
+	lf := map[string]interface{}{"encJWT": encJWT}
+	ctx.WithFields(lf).Debug("validating jwt")
+
+	jwt, err := jws.ParseJWT([]byte(encJWT))
+	if err != nil {
+		ctx.WithFields(lf).WithError(err).Error("error parsing jwt")
+		return nil, &types.ErrSecTokInvalid{InvalidToken: true, InnerError: err}
+	}
+
+	sm := parseSigningMethod(config.Alg)
+	lf["signingMethod"] = sm.Alg()
+	ctx.WithFields(lf).Debug("parsed jwt signing method")
+
+	if err := jwt.Validate(config.Key, sm); err != nil {
+		ctx.WithFields(lf).WithError(err).Error("error validating jwt")
+		return nil, &types.ErrSecTokInvalid{InvalidSig: true, InnerError: err}
+	}
+
+	ctx.WithFields(lf).Debug("validated jwt signature")
+
+	if len(jwt.Claims()) == 0 {
+		ctx.WithFields(lf).Error("jwt missing claims")
+		return nil, &types.ErrSecTokInvalid{}
+	}
+
+	var (
+		sub string
+		iat time.Time
+		exp time.Time
+		nbf time.Time
+		ok  bool
+	)
+
+	if sub, ok = jwt.Claims().Subject(); !ok {
+		ctx.WithFields(lf).Error("jwt missing sub claim")
+		return nil, &types.ErrSecTokInvalid{MissingClaim: "sub"}
+	}
+
+	if iat, ok = jwt.Claims().IssuedAt(); !ok {
+		ctx.WithFields(lf).Error("jwt missing iat claim")
+		return nil, &types.ErrSecTokInvalid{MissingClaim: "iat"}
+	}
+
+	if exp, ok = jwt.Claims().Expiration(); !ok {
+		ctx.WithFields(lf).Error("jwt missing exp claim")
+		return nil, &types.ErrSecTokInvalid{MissingClaim: "exp"}
+	}
+
+	if nbf, ok = jwt.Claims().NotBefore(); !ok {
+		ctx.WithFields(lf).Error("jwt missing nbf claim")
+		return nil, &types.ErrSecTokInvalid{MissingClaim: "nbf"}
+	}
+
+	tok := &types.SecurityToken{
+		Subject:   sub,
+		IssuedAt:  iat.UTC().Unix(),
+		Expires:   exp.UTC().Unix(),
+		NotBefore: nbf.UTC().Unix(),
+	}
+
+	lf["sub"] = tok.Subject
+	lf["iat"] = tok.IssuedAt
+	lf["exp"] = tok.Expires
+	lf["nbf"] = tok.NotBefore
+
+	var (
+		allowed bool
+		denied  bool
+	)
+	for _, v := range config.Deny {
```

Comment:
This code that loops through Denied and Allow subjects identical to that found in the service auth handler (when a previously validated token is already in the context). Could we move this bit of logic in a common function, such as `security.ValidateTokenSubject(config, tok.Subject)` and call that from both places?